File: packages/core/src/semantics/pickFirstRelease.ts
import type { SortKey, TagInfo } from '../types.js';
import { matchesGlob } from './glob.js';

export interface PickFirstReleaseArgs {
  /** Tag names returned by `git tag --contains`. */
  containingTagNames: readonly string[];
  /** Metadata index for every tag we care about (typically all repo tags). */
  tagIndex: ReadonlyMap<string, TagInfo>;
  /** Glob patterns to drop (e.g. `*-rc*`). */
  exclude?: readonly string[];
  /** Order key. v0.1 only implements `committerDate`; others fall back to it. */
  sortBy?: SortKey;
}

export interface PickFirstReleaseResult {
  picked: TagInfo | null;
  /** All candidates left after filtering, sorted ascending by the chosen key. */
  candidates: TagInfo[];
}

/**
 * Apply ShipLens' "first release" rules to the set of containing tags. The
 * picking logic is pure and isolated so it can be unit-tested without touching
 * git.
 */
export function pickFirstRelease(args: PickFirstReleaseArgs): PickFirstReleaseResult {
  const { containingTagNames, tagIndex, exclude = [], sortBy = 'committerDate' } = args;

  const candidates: TagInfo[] = [];
  for (const name of containingTagNames) {
    const meta = tagIndex.get(name);
    if (!meta) continue;
    if (exclude.some((pat) => matchesGlob(name, pat))) continue;
    candidates.push(meta);
  }

  candidates.sort((a, b) => compare(a, b, sortBy));
  return { picked: candidates[0] ?? null, candidates };
}

function compare(a: TagInfo, b: TagInfo, key: SortKey): number {
  switch (key) {
    case 'committerDate':
    case 'tagDate':
    case 'topological':
    case 'semver':
    default:
      // v0.1: only committerDate is implemented. tagDate / topological /
      // semver fall through to the same comparison and are scheduled for
      // a later iteration.
      return a.committerDate.localeCompare(b.committerDate);
  }
}

File: packages/jetbrains-plugin/src/main/kotlin/com/shiplens/jetbrains/core/semantics/PickFirstRelease.kt
package com.shiplens.jetbrains.core.semantics

import com.shiplens.jetbrains.core.SortKey
import com.shiplens.jetbrains.core.TagInfo

internal data class PickFirstReleaseResult(
  val picked: TagInfo?,
  /** All candidates left after filtering, sorted ascending by the chosen key. */
  val candidates: List<TagInfo>,
)

/**
 * Apply ShipLens' "first release" rules to the set of containing tags.
 * Pure logic, mirror of @shiplens/core's `semantics/pickFirstRelease.ts`.
 */
internal fun pickFirstRelease(
  containingTagNames: List<String>,
  tagIndex: Map<String, TagInfo>,
  exclude: List<String> = emptyList(),
  sortBy: SortKey = SortKey.COMMITTER_DATE,
): PickFirstReleaseResult {
  val candidates = containingTagNames
    .asSequence()
    .mapNotNull { tagIndex[it] }
    .filter { meta -> exclude.none { pat -> Glob.matches(meta.name, pat) } }
    .sortedWith(compare(sortBy))
    .toList()

  return PickFirstReleaseResult(picked = candidates.firstOrNull(), candidates = candidates)
}

private fun compare(@Suppress("UNUSED_PARAMETER") sortBy: SortKey): Comparator<TagInfo> {
  // v0.1: only committerDate is implemented. Other keys fall through to the
  // same comparison and are tracked for a later iteration.
  return compareBy { it.committerDate }
}
